### ISSUE TITLE
Segfault in PyObjCBlock_GetSignature (do Python objects need to inherit from NSObject?)

### ISSUE THREAD

**Issue by mrpippy** [ronaldoussoren/pyobjc#594]
**Describe the bug**

Our application is Objective-C, but uses PyObjC to call into cross-platform Python code.

We are seeing a very sporadic segfault in PyObjC, when calling from ObjC into Python and passing a Python object as an argument. After disabling LTO and building our app with debug symbols, this lldb snippet shows it:

```
(lldb) cont
Process 35413 resuming
Process 35413 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x657875)
    frame #0: 0x000000011b787b65 _objc.cpython-311-darwin.so`PyObjCBlock_GetSignature(_block=0x000000011fe72f90) at block_support.m:79:16 [opt]
   76           if (block->flags & BLOCK_HAS_COPY_DISPOSE) {
   77               signature_loc += sizeof(void (*)(void)) * 2;
   78           }
-> 79           return *(const char**)signature_loc;
   80       }
   81       return NULL;
   82   }
Target 0: (CrossOver) stopped.
warning: _objc.cpython-311-darwin.so was compiled with optimization - stepping may behave oddly; variables may not be available.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x657875)
  * frame #0: 0x000000011b787b65 _objc.cpython-311-darwin.so`PyObjCBlock_GetSignature(_block=0x000000011fe72f90) at block_support.m:79:16 [opt]
    frame #1: 0x000000011b7bbfab _objc.cpython-311-darwin.so`method_stub(cif=<unavailable>, resp=0x00007ff7b1a13db0, args=0x00007ff7b1a13c40, _userdata=0x000000011c0fc190) at libffi_support.m:1413:45 [opt]
    frame #2: 0x00007ff827bc54b1 libffi.dylib`ffi_closure_unix64_inner + 510
    frame #3: 0x00007ff827bc5a2c libffi.dylib`ffi_closure_unix64 + 72
    frame #4: 0x000000010e4f19a9 CrossOver`-[PackageInstaller addTaskToQueue:](self=0x0000600000208960, _cmd="addTaskToQueue:", task=0x0000600003006640) at PackageInstaller.m:314:10
    frame #5: 0x000000010e4f24ee CrossOver`-[PackageInstaller spoolAvailableTasks](self=0x0000600000208960, _cmd="spoolAvailableTasks") at PackageInstaller.m:417:9
    frame #6: 0x000000010e4f3178 CrossOver`-[PackageInstaller continueInstallation](self=0x0000600000208960, _cmd="continueInstallation") at PackageInstaller.m:516:9
    frame #7: 0x000000010e4f28f2 CrossOver`-[PackageInstaller taskCompleted:](self=0x0000600000208960, _cmd="taskCompleted:", taskStatus=0x000060000bc867f0) at PackageInstaller.m:449:5
    frame #8: 0x00007ff8177dbb65 Foundation`__NSThreadPerformPerform + 178
...
(lldb) print block
(block_literal *) 0x000000011fe72f90
(lldb) print *block
(block_literal) {
  isa = 0x0000000000000008
  flags = 1222149120
  reserved = 32726
  invoke = 0x0000000000000000
  descriptor = 0x0000000000657865
  invoke_cleanup = nil
}
```

Clearly something is wrong here: `reserved` should be 0. `descriptor` is invalid, and a NULL `invoke` doesn't make sense either. Also, we're not passing blocks in/out of Python code at all.

On another run, in `method_stub` I printed `v` as a `PyObjCObject`:
```
(lldb) print *(PyObjCObject*)v
(PyObjCObject) {
  ob_base = {
    ob_refcnt = 8
    ob_type = 0x00007feb7b8d0770
  }
  objc_object = nil
  flags = 7565925
}
```

That also doesn't look right, at the very least there are many unassigned flag bits set.

I then realized that this object doesn't inherit from `Foundation.NSObject`. It's a plain Python class, returned to ObjC by a Python function, then the ObjC code is passing it back into Python as the argument to a function.

Maybe the problem is that the PyObjC-part of this object is never being initialized correctly, so it ends up with uninitialized data? And that usually works for some reason, but if `flags` happens to include `PyObjCObject_kBLOCK`, then we end up in `PyObjCBlock_GetSignature()` and crash?

When running with `PYTHONMALLOC="debug"`, I see `block->descriptor` full of `0xfd`, which is the "forbidden byte".

**Platform information**

* Python 3.11.8, built from source as a framework
* Binary is x86_64, I haven't tried to reproduce with arm64.
* macOS 14.3.1

**To Reproduce**
I've attached a standalone test case, for use with Python 3.11.8 and the latest PyObjC from git.
With unmodified PyObjC it'll likely crash very infrequently, but by adding an assert to `PyObjCBlock_GetSignature()` to check `reserved` (see [1]) I usually see the problematic codepath within a couple runs. Attaching lldb (`lldb -w -n main`, then set a breakpoint on `abort()`) works well.

[main.m.txt](https://github.com/ronaldoussoren/pyobjc/files/14390862/main.m.txt)
[main.py.txt](https://github.com/ronaldoussoren/pyobjc/files/14390864/main.py.txt)
(you'll need to rename the files to remove `.txt`)

My main question: is the bug here that the Python `Task` class doesn't inherit from `Foundation.NSObject`? It isn't really documented.
If so, we can fix that. But it would be nice to have warnings/errors when a non-NSObject is being passed to ObjC code.
At the very least, some asserts should be added in `PyObjCBlock_GetSignature()` to check `reserved`, and maybe in `PyObjCObject_GetFlags()` to check that only defined flag bits are set.

[1]:
```
diff --git a/pyobjc-core/Modules/objc/block_support.m b/pyobjc-core/Modules/objc/block_support.m
index 664951bf7..fc668030b 100644
--- a/pyobjc-core/Modules/objc/block_support.m
+++ b/pyobjc-core/Modules/objc/block_support.m
@@ -59,6 +59,9 @@
 {
     struct block_literal* block = (struct block_literal*)_block;
 
+    if (block->reserved != 0)
+           abort();
+
     if (((intptr_t)(block->isa)) & 0x1) {
         /* XXX: Hack to avoid a hard crash that I haven't been able
          * to pintpoint yet (test doesn't fail reliably, and I haven't
```

**Comment by mrpippy** [comment]
Hi @ronaldoussoren, do you have any thoughts on this? Should our class (`Task` in the example) be inheriting from NSObject?

**Comment by ronaldoussoren** [comment]
This might be a reference counting issue "somewhere".

> I then realized that this object doesn't inherit from Foundation.NSObject. It's a plain Python class, returned to ObjC by a Python function, then the ObjC code is passing it back into Python as the argument to a function.

Is there an autorelease pool that gets recycled between returning the Python instance and passing it back into Python?

In general Python instances get proxied into an instance of an Objective-C class (the various `OC_` classes in pyobjc-core), and those instances are autoreleased to conform to the Objective-C calling conventions.  This means that you can end up with an invalid pointer when Objective-C's autorelease pool is recycled (assuming manual refcounting, ARC should do the right thing automatically). 

For instances of subclasses of Foundation.NSObject the "real" value of the instance is the Objective-C representation, and the Python value of the instance is a proxy. That's also true for subclasses implemented in Python.  


Looking at your attached code:

```python
class TaskEngine(Foundation.NSObject):
    def __init__(self):
        self._tasks = set()
        for x in range(100):
            self._tasks.add(Task())
```

Please don't do this, but perform initialisation using the normal Objective-C pattern. This can cause problems when an instance of this class is kept alive in Objective-C and then passed back into Python. That's not the root cause for the crash though.

Annoyingly I haven't yet been able to reproduce this with a regular framework install (python.org installer) and PyObjC build from source, even with the additional assertion you mention.

**Comment by mrpippy** [comment]
> This might be a reference counting issue "somewhere".
> 
> > I then realized that this object doesn't inherit from Foundation.NSObject. It's a plain Python class, returned to ObjC by a Python function, then the ObjC code is passing it back into Python as the argument to a function.
> 
> Is there an autorelease pool that gets recycled between returning the Python instance and passing it back into Python?

No, as in the test `main.m` there's no autorelease pool draining. ARC is being used. In `test.m` I even see crashes if I take out the `for (Task *task in engine.returnTasksList) [task func];` block, so all the Tasks are proxied, sent immediately back to Python, and that's when it (sometimes) crashes.

> In general Python instances get proxied into an instance of an Objective-C class (the various `OC_` classes in pyobjc-core), and those instances are autoreleased to conform to the Objective-C calling conventions. This means that you can end up with an invalid pointer when Objective-C's autorelease pool is recycled (assuming manual refcounting, ARC should do the right thing automatically).
> 
> For instances of subclasses of Foundation.NSObject the "real" value of the instance is the Objective-C representation, and the Python value of the instance is a proxy. That's also true for subclasses implemented in Python.
> 
> Looking at your attached code:
> 
> ```python
> class TaskEngine(Foundation.NSObject):
>     def __init__(self):
>         self._tasks = set()
>         for x in range(100):
>             self._tasks.add(Task())
> ```
> 
> Please don't do this, but perform initialisation using the normal Objective-C pattern. This can cause problems when an instance of this class is kept alive in Objective-C and then passed back into Python. That's not the root cause for the crash though.

I believe `__init__` only gets called from the `initFromObjC` function, which does call `super().init()` first, should that work correctly?

(I'm copying the structure of our internal code, these classes are used both on Linux and through PyObjC, we do as much initialization as possible in `__init__` so they're usable in the normal Python way)

> Annoyingly I haven't yet been able to reproduce this with a regular framework install (python.org installer) and PyObjC build from source, even with the additional assertion you mention.

I tried this with the Python 3.11.8 pkg installer and binary PyObjC from pip, and with Guard Malloc+`PYTHONMALLOC=malloc_debug` I consistently get a segfault (with x86_64 or arm):

```
$ /Library/Frameworks/Python.framework/Versions/Current/bin/pip3 install pyobjc-core
$ /Library/Frameworks/Python.framework/Versions/Current/bin/pip3 install pyobjc-framework-Cocoa
$ clang -o bin/main main.m -fobjc-arc -framework Foundation -framework Python -F /Library/Frameworks
** check the python version: **
$ otool -L bin/main
bin/main:
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 2420.0.0)
	/Library/Frameworks/Python.framework/Versions/3.11/Python (compatibility version 3.11.0, current version 3.11.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.100.2)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 2420.0.0)
$ lldb bin/main
(lldb) target create "bin/main"
Current executable set to '/Users/pip/bug-pocs/pyobjc/bin/main' (arm64).
(lldb) settings set target.env-vars DYLD_INSERT_LIBRARIES=/usr/lib/libgmalloc.dylib
(lldb) settings set target.env-vars PYTHONMALLOC="malloc_debug"
(lldb) run
Process 55528 launched: '/Users/pip/bug-pocs/pyobjc/bin/main' (arm64)
GuardMalloc[main-55528]: Allocations will be placed on 16 byte boundaries.
GuardMalloc[main-55528]:  - Some buffer overruns may not be noticed.
GuardMalloc[main-55528]:  - Applications using vector instructions (e.g., SSE) should work.
GuardMalloc[main-55528]: version 064562.4.1.1
PyRun ret 0
Process 55528 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x2c12b0000)
    frame #0: 0x000000011684f0fc _objc.cpython-311-darwin.so`method_stub.llvm.17679664296054037312 + 696
_objc.cpython-311-darwin.so`method_stub.llvm.17679664296054037312:
->  0x11684f0fc <+696>: ldr    x8, [x27, #0x20]
    0x11684f100 <+700>: cbnz   x8, 0x11684f230           ; <+1004>
    0x11684f104 <+704>: ldr    x8, [x22]
    0x11684f108 <+708>: ldr    x8, [x8, #0x8]
Target 0: (main) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x2c12b0000)
  * frame #0: 0x000000011684f0fc _objc.cpython-311-darwin.so`method_stub.llvm.17679664296054037312 + 696
    frame #1: 0x000000019ab9ff28 libffi.dylib`ffi_closure_SYSV_inner + 816
    frame #2: 0x000000019ab971e8 libffi.dylib`ffi_closure_SYSV + 56
    frame #3: 0x0000000100003c94 main`main + 528
    frame #4: 0x0000000189c810e0 dyld`start + 2360
```

But crucially, I also tried the binary Python 3.12.2 with the same options and don't see a segfault. This crash did only appear once we upgraded to Python 3.11, maybe it's a 3.11-specific bug.

**Comment by mrpippy** [comment]
It's not a very satisfying conclusion, but this bug seems to be specific to Python 3.11. I can't reproduce it with 3.10 or 3.12. Our "fix" here is to just downgrade to 3.10 or upgrade to 3.12.
